src/Compilers/Core/Portable/InternalUtilities/OrderedMultiDictionary.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Roslyn.Utilities
+{
+    // Note that this is not threadsafe for concurrent reading and writing.
+    internal sealed class OrderedMultiDictionary<K, V> : IEnumerable<KeyValuePair<K, SetWithInsertionOrder<V>>>
+    {
+        private readonly Dictionary<K, SetWithInsertionOrder<V>> _dictionary;
+        private readonly List<K> _keys;
+
+        public int Count => _dictionary.Count;
+
+        public IEnumerable<K> Keys => _keys;
+
+        // Returns an empty set if there is no such key in the dictionary.
+        public SetWithInsertionOrder<V> this[K k]
+        {
+            get
+            {
+                SetWithInsertionOrder<V> set;
+                return _dictionary.TryGetValue(k, out set)
+                    ? set : new SetWithInsertionOrder<V>();
```

Comment:
@cston I would like to, but that breaks the expected semantics of the original multidictionary, which explicitly chose this path for some reason. I think we should change this behavior, but I think that is out of the scope of this PR.